### PR TITLE
Fix performance interface timing member

### LINKED ISSUE #23330: Performance interface's timing member should be an attribute
Per https://w3c.github.io/navigation-timing/#extensions-to-the-performance-interface we should have a `timing` attribute in our WebIDL interface, not a method.

Code: `components/script/dom/webidls/Performance.webidl`, `components/script/dom/performance.rs`
Tests: `./mach test-wpt tests/wpt/web-platform-tests/navigation-timing/`

### PR DESCRIPTION
**Pull request by shnmorimoto**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #23330 and #24468 (GitHub issue number if applicable)

- [x] There are tests for these changes

### CODE REVIEW

**Comment by shnmorimoto** [comment]
When I run `./mach run https://www.google.com `, I get a crash.

```
> ./mach run https://www.google.com                                                                                                                                    fix_performance_interface_timing_member ✱
[2019-12-08T07:06:36Z ERROR script::dom::bindings::error] Error at https://www.google.com/:345:1419 document.f.q is undefined
internal error: entered unreachable code: Are we trying to expose Performance.timing in workers? (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at components/script/dom/
performance.rs:394)
stack backtrace:
   0: backtrace::backtrace::trace_unsynchronized
   1: <servo::backtrace::Print as core::fmt::Debug>::fmt
   2: core::fmt::write
   3: <std::io::stdio::Stdout as std::io::Write>::write_fmt
   4: std::io::stdio::_print
   5: servo::backtrace::print
   6: servo::main::{{closure}}
   7: std::panicking::rust_panic_with_hook
   8: rust_begin_unwind
   9: std::panicking::begin_panic_fmt
  10: <script::dom::performance::Performance as script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::PerformanceMethods>::Timing
  11: script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::get_timing::{{closure}}
  12: core::ops::function::FnOnce::call_once
  13: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  14: std::panicking::try::do_call
  15: __rust_maybe_catch_panic
  16: std::panicking::try
  17: std::panic::catch_unwind
  18: mozjs::panic::wrap_panic
  19: script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::get_timing
  20: CallJitGetterOp
  21: script::dom::bindings::utils::generic_call
  22: script::dom::bindings::utils::generic_getter
  23: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructE
  24: _ZN2js10CallGetterEP9JSContextN2JS6HandleINS2_5ValueEEES5_NS2_13MutableHandleIS4_EE
  25: _ZN2js17NativeGetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_INS2_5ValueEEENS3_INS2_11PropertyKeyEEENS2_13MutableHandleIS7_EE
  26: _ZN2js3jit17DoGetElemFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_18ICGetElem_FallbackEN2JS6HandleINS7_5ValueEEESA_NS7_13MutableHandleIS9_EE

[2019-12-08T07:06:38Z ERROR servo] internal error: entered unreachable code: Are we trying to expose Performance.timing in workers?
```

I'm trying to find out the cause of this error.

**Comment by shnmorimoto** [comment]
@jdm 
I inserted debug log like this. 

```rust
impl PerformanceMethods for Performance {
    // FIXME(avada): this should be deprecated in the future, but some sites still use it
    // https://dvcs.w3.org/hg/webperf/raw-file/tip/specs/NavigationTiming/Overview.html#performance-timing-attribute
    fn Timing(&self) -> DomRoot<PerformanceNavigationTiming> {
        debug!("#25205 call Timing function");
        let entries = self.GetEntriesByType(DOMString::from("navigation"));
 ....
```

And,  I investigated when each function was called.

```
[2019-12-09T15:01:15Z DEBUG script::dom::servoparser] #25205 call process_response func
[2019-12-09T15:01:15Z DEBUG script::dom::servoparser] #25205 call submit_resource_timing func
[2019-12-09T15:01:16Z DEBUG script::dom::performance] #25205 call Timing function
[2019-12-09T15:01:20Z DEBUG script::dom::servoparser] #25205 call process_response_eof func
[2019-12-09T15:01:20Z DEBUG script::dom::performance] #25205 call update_entry function
[2019-12-09T15:01:21Z DEBUG script::dom::performance] #25205 call Timing function
[2019-12-09T15:01:21Z DEBUG script::dom::performance] #25205 call Timing function
internal error: entered unreachable code: Are we trying to expose Performance.timing in workers? (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at components/script/dom/performance.rs:396)
[2019-12-09T15:01:21Z ERROR servo] internal error: entered unreachable code: Are we trying to expose Performance.timing in workers?
[2019-12-09T15:01:22Z DEBUG constellation::constellation] Sending log entry Panic("internal error: entered unreachable code: Are we trying to expose Performance.timing in workers?", "stack backtrace:\n   0: backtrace::backtrace::trace_unsynchronized\n   1: backtrace::backtrace::trace\n   2: backtrace::capture::Backtrace::create\n   3: backtrace::capture::Backtrace::new\n   4: constellation::constellation::log_entry\n   5: <constellation::constellation::FromCompositorLogger as log::Log>::log\n   6: <servo::BothLogger<Log1,Log2> as log::Log>::log\n   7: log::__private_api_log\n   8: servo::main::{{closure}}\n   9: std::panicking::rust_panic_with_hook\n  10: rust_begin_unwind\n  11: std::panicking::begin_panic_fmt\n  12: <script::dom::performance::Performance as script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::PerformanceMethods>::Timing\n  13: script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::get_timing::{{closure}}\n  14: core::ops::function::FnOnce::call_once\n  15: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n  16: std::panicking::try::do_call\n  17: __rust_maybe_catch_panic\n  18: std::panicking::try\n  19: std::panic::catch_unwind\n  20: mozjs::panic::wrap_panic\n  21: script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::get_timing\n  22: CallJitGetterOp\n  23: script::dom::bindings::utils::generic_call\n  24: script::dom::bindings::utils::generic_getter\n  25: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructE\n  26: _ZN2js10CallGetterEP9JSContextN2JS6HandleINS2_5ValueEEES5_NS2_13MutableHandleIS4_EE\n  27: _ZN2js17NativeGetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_INS2_5ValueEEENS3_INS2_11PropertyKeyEEENS2_13MutableHandleIS7_EE\n  28: _ZN2js3jit17DoGetElemFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_18ICGetElem_FallbackEN2JS6HandleINS7_5ValueEEESA_NS7_13MutableHandleIS9_EE\n").
[2019-12-09T15:01:22Z DEBUG constellation::constellation] Received log entry Panic("internal error: entered unreachable code: Are we trying to expose Performance.timing in workers?", "stack backtrace:\n   0: backtrace::backtrace::trace_unsynchronized\n   1: backtrace::backtrace::trace\n   2: backtrace::capture::Backtrace::create\n   3: backtrace::capture::Backtrace::new\n   4: constellation::constellation::log_entry\n   5: <constellation::constellation::FromCompositorLogger as log::Log>::log\n   6: <servo::BothLogger<Log1,Log2> as log::Log>::log\n   7: log::__private_api_log\n   8: servo::main::{{closure}}\n   9: std::panicking::rust_panic_with_hook\n  10: rust_begin_unwind\n  11: std::panicking::begin_panic_fmt\n  12: <script::dom::performance::Performance as script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::PerformanceMethods>::Timing\n  13: script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::get_timing::{{closure}}\n  14: core::ops::function::FnOnce::call_once\n  15: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n  16: std::panicking::try::do_call\n  17: __rust_maybe_catch_panic\n  18: std::panicking::try\n  19: std::panic::catch_unwind\n  20: mozjs::panic::wrap_panic\n  21: script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::get_timing\n  22: CallJitGetterOp\n  23: script::dom::bindings::utils::generic_call\n  24: script::dom::bindings::utils::generic_getter\n  25: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructE\n  26: _ZN2js10CallGetterEP9JSContextN2JS6HandleINS2_5ValueEEES5_NS2_13MutableHandleIS4_EE\n  27: _ZN2js17NativeGetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_INS2_5ValueEEENS3_INS2_11PropertyKeyEEENS2_13MutableHandleIS7_EE\n  28: _ZN2js3jit17DoGetElemFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_18ICGetElem_FallbackEN2JS6HandleINS7_5ValueEEESA_NS7_13MutableHandleIS9_EE\n").
[2019-12-09T15:01:22Z DEBUG constellation::constellation] Panic handler for top-level browsing context (0,1): internal error: entered unreachable code: Are we trying to expose Performance.timing in workers?.
[2019-12-09T15:01:22Z DEBUG script::dom::servoparser] #25205 call process_response func
[2019-12-09T15:01:22Z DEBUG script::dom::servoparser] #25205 call submit_resource_timing func
[2019-12-09T15:01:22Z DEBUG script::dom::servoparser] #25205 call process_response_eof func
[2019-12-09T15:01:22Z DEBUG script::dom::performance] #25205 call update_entry function
```

From the above results, it seems that there is a timing when `Timing` func is called earlier than `process_response` func.

Is there a way to investigate the cause of the problem?

**Comment by jdm** [comment]
Yes, this is https://github.com/servo/servo/issues/24468. I believe I have described the solution I have in mind that in that issue; let me know if it is unclear at all!

**Comment by jdm** [comment]
Oh, sorry, I didn't read your last message closely enough. I'll try and figure out what to do.

**Comment by jdm** [comment]
@shnmorimoto Do either of the following trigger the problem?
```html
<iframe src="javascript:console.log(navigation.timing)"></iframe>
```
```
<iframe srcdoc="<script>console.log(navigation.timing)</script>"></iframe>
```

**Comment by shnmorimoto** [comment]
@jdm Thanks! Neither way caused problems.
This is the html used for the test.

```html
<!DOCTYPE html>
<html>
<head>
<title>iframe 1</title>
</head>
<body>

Hello World.
<iframe src="javascript:console.log(performance.timing)"></iframe>

</body>
</html>
```

```html
<!DOCTYPE html>
<html>
<head>
<title>iframe 2</title>
</head>
<body>

Hello World.

<iframe srcdoc="<script>console.log(performance.timing)</script>"></iframe>

</body>
</html>
```

**Comment by jdm** [comment]
@shnmorimoto I ended up using a debugger to figure out what was going on - I set a breakpoint in `Performance::queue_entry` and watched as `self->buffer` grew over time, then suddenly was empty. Looking at the code that modifies `self.buffer`, I realized that `clear_entries_by_name_and_type` has a bug - it is modeled after `get_entries_by_name_and_type`, so it ends up keeping any entries that match the name/type being cleared and throwing away any non-matching ones, which is the opposite behaviour from what we want. Fixing that bug should fix the google.com problem.

**Comment by shnmorimoto** [comment]
@jdm Thank you for finding the cause and explaining how to way to find out.
I will try it when the same thing happens next time.

I understand the cause. I'll try to fix it.

**Comment by shnmorimoto** [comment]
I've fixed the `clear_entries_by_name_and_type` bug and confirmed that the google.com problem is resolved.

File: components/script/dom/performance.rs [hunk 1]
```diff
@@ -96,10 +96,10 @@ impl PerformanceEntryList {
         entry_type: Option<DOMString>,
     ) {
         self.entries.retain(|e| {
-            name.as_ref().map_or(true, |name_| *e.name() == *name_) &&
+            name.as_ref().map_or(true, |name_| *e.name() != *name_) ||
```
Comment (jdm) [inline components/script/dom/performance.rs]: I think we either need to use map_or(false, ...) with ||, or map_or(true, ...) with &&. The current code means that if there is no name passed but the entry type matches then we will not clear the entry even though we should. Similarly, if there is no entry passed but the names match, we will ignore it.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/performance.rs (hunk 1 above)
Comment (shnmorimoto) [reply components/script/dom/performance.rs]: Thank you for the review. Oh, I misunderstood the clear condition.
Fixed it.

**Comment by jdm** [comment]
Lots of improvements, and one worker test where the panic about exposing Performance.timing in a worker is actually reporting the right thing for once.
```
  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  └ PASS [expected FAIL] Can use clearMarks and clearMeasures in workers
  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  └ PASS [expected FAIL] performance.clearResourceTimings in workers

  ▶ Unexpected subtest result in /resource-timing/clear_resource_timing_functionality.html:
  └ PASS [expected FAIL] No resource timing entries should be stored after clearResourceTimings.
  ▶ OK [expected TIMEOUT] /performance-timeline/webtiming-resolution.any.html
  ▶ Unexpected subtest result in /performance-timeline/webtiming-resolution.any.html:
  │ FAIL [expected PASS] Verifies the resolution of performance.now() is at least 5 microseconds.
  │   → assert_greater_than_equal: The second performance.now() should be much greater than the first expected a number greater than or equal to 0.0049900000000000005 but got 0.0015530000000012478
  │ 
  │ testTimeResolution/<@http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.js:9:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1984:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ testTimeResolution@http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.js:2:5
  └ @http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.js:24:1

  â–¶ Unexpected subtest result in /hr-time/performance-tojson.html:
  â”” PASS [expected FAIL] Test performance.toJSON()


  â–¶ CRASH [expected OK] /hr-time/idlharness.any.worker.html
  â”‚ 
  â”‚ 
  â”‚ 
  â”‚ 
  â”‚ internal error: entered unreachable code: Are we trying to expose Performance.timing in workers? (thread WebWorker for http://web-platform.test:8000/hr-time/idlharness.any.worker.js, at components/script/dom/performance.rs:394)
  â”‚ stack backtrace:
  â”‚    0: servo::backtrace::print
  â”‚    1: servo::main::{{closure}}
  â”‚    2: std::panicking::rust_panic_with_hook
  â”‚              at src/libstd/panicking.rs:475
  â”‚    3: rust_begin_unwind
  â”‚              at src/libstd/panicking.rs:375
  â”‚    4: std::panicking::begin_panic_fmt
  â”‚              at src/libstd/panicking.rs:326
  â”‚    5: <script::dom::performance::Performance as script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::PerformanceMethods>::Timing
  â”‚    6: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  â”‚    7: _ZN3std9panicking3try7do_call17hf46efd93f47ab473E.llvm.1649388449213437728
  â”‚    8: __rust_maybe_catch_panic
  â”‚              at src/libpanic_unwind/lib.rs:78
  â”‚    9: script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::CollectJSONAttributes
  â”‚   10: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  â”‚   11: _ZN3std9panicking3try7do_call17h7de1923c0cb16390E.llvm.1649388449213437728
  â”‚   12: __rust_maybe_catch_panic
  â”‚              at src/libpanic_unwind/lib.rs:78
  â”‚   13: script::dom::bindings::codegen::Bindings::PerformanceBinding::PerformanceBinding::toJSON
  â”‚   14: CallJitMethodOp
  â”‚   15: script::dom::bindings::utils::generic_call
  â”‚   16: _Z12CallJSNativeP9JSContextPFbS0_jPN2JS5ValueEERKNS1_8CallArgsE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/vm/Interpreter.cpp:442
  â”‚       _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/vm/Interpreter.cpp:534
  â”‚   17: _ZN2js13CallFromStackEP9JSContextRKN2JS8CallArgsE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/vm/Interpreter.cpp:593
  â”‚       _ZL9InterpretP9JSContextRN2js8RunStateE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/vm/Interpreter.cpp:3075
  â”‚   18: _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/vm/Interpreter.cpp:422
  â”‚   19: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/vm/Interpreter.cpp:562
  â”‚   20: _ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/vm/Interpreter.cpp:605
  â”‚   21: _ZN2js3jit14InvokeFunctionEP9JSContextN2JS6HandleIP8JSObjectEEbbjPNS3_5ValueENS3_13MutableHandleIS8_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/jit/VMFunctions.cpp:260
  â”‚   22: _ZN2js3jit25InvokeFromInterpreterStubEP9JSContextPNS0_30InterpreterStubExitFrameLayoutE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/jit/VMFunctions.cpp:289
  â”‚   23: <unknown>
  â”‚   24: <unknown>
  â”‚   25: <unknown>
  â”‚   26: <unknown>
  â”‚   27: <unknown>
  â”‚   28: <unknown>
  â”‚   29: <unknown>
  â”‚   30: <unknown>
  â”‚   31: <unknown>
  â”‚   32: _ZL8EnterJitP9JSContextRN2js8RunStateEPh
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/jit/Jit.cpp:103
  â”‚       _ZN2js3jit13MaybeEnterJitEP9JSContextRNS_8RunStateE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/5906588/mozjs/js/src/jit/Jit.cpp:168
  â”‚   33: <unknown>
  â”‚ 
  â”‚ [2019-12-11T16:05:47Z ERROR servo] internal error: entered unreachable code: Are we trying to expose Performance.timing in workers?
  â”” Pipeline failed in hard-fail mode.  Crashing!
```

**Comment by jdm** [comment]
Let's go ahead and mark all those test result changes as expected and file a followup issue for the panic in /hr-time/idlharness.any.worker.html.

**Comment by shnmorimoto** [comment]
@jdm I know to update the test if unexpected result is OK and fix the panic. But What should I do if unexpected is Error or Timeout, NOTRUN?

**Comment by jdm** [comment]
`expected: ERROR`, `expected: TIMEOUT`, `expected: NOTRUN` :)

**Comment by shnmorimoto** [comment]
@jdm Thanks! 
I will run `/mach test-wpt tests/wpt/web-platform-tests/xxxxx --log-raw /tmp/servo.log ` and `./mach update-wpt /tmp/servo.log `.

**Comment by shnmorimoto** [comment]
I've marked tests and filed an issue about crash.
https://github.com/servo/servo/issues/25281

**Comment by shnmorimoto** [comment]
Oops, Sorry, I will fix.

**Comment by jdm** [comment]
To be clear, these changes are necessary:
```
  ▶ ERROR [expected TIMEOUT] /performance-timeline/multiple-buffered-flag-observers.any.worker.html
  └   → Error in remote http://web-platform.test:8000/performance-timeline/multiple-buffered-flag-observers.any.js: Missing required member "entryTypes".
  ▶ ERROR [expected TIMEOUT] /performance-timeline/multiple-buffered-flag-observers.any.html
  │   → Missing required member "entryTypes".
  │ 
  │ promise2</<@http://web-platform.test:8000/performance-timeline/multiple-buffered-flag-observers.any.js:25:10
  └ step_timeout/<@http://web-platform.test:8000/resources/testharness.js:894:15
  ▶ Unexpected subtest result in /performance-timeline/webtiming-resolution.any.html:
  │ FAIL [expected PASS] Verifies the resolution of performance.now() is at least 5 microseconds.
  │   → assert_greater_than_equal: The second performance.now() should be much greater than the first expected a number greater than or equal to 0.0049900000000000005 but got 0.003284000000007836
  │ 
  │ testTimeResolution/<@http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.js:9:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1984:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ testTimeResolution@http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.js:2:5
  └ @http://web-platform.test:8000/performance-timeline/webtiming-resolution.any.js:24:1

  ▶ OK [expected TIMEOUT] /_webgl/conformance/glsl/bugs/nested-functions-should-not-crash.html
```

**Comment by shnmorimoto** [comment]
Thanks! I didn't reproduce it locally, so I changed it manually.

**Comment by jdm** [comment]
@bors-servo retry
* #24762

**Comment by jdm** [comment]
@bors-servo retry
* #25296 #24622

**Pull request merged**